### ISSUE TITLE
[JENKINS-41205] Stage graph unsuitable for large and/or complex pipelines

### ISSUE THREAD

**Issue by jenkins-infra-bot** [jenkinsci/blueocean-plugin#3641]
<div class="panel" style="border-width: 1px;"><div class="panelHeader" style="border-bottom-width: 1px;background-color: #3878de;color: white;"><b>Improvement on roadmap</b></div><div class="panelContent">
<p>This improvement is on the Blue Ocean project roadmap. Check the <a href="https://jenkins.io/projects/blueocean/roadmap/" class="external-link" target="_blank" rel="nofollow noopener">roadmap page</a> for updates. </p>
</div></div>

<p>The Blue Ocean stage graph is great for small, simple pipelines however it breaks down with many parallel builds. See attached screenshot for an example.</p>

<p>Because 'stage' can no longer be nested within 'parallel', all of our steps must belong under a single 'Test' stage. We have 19 parallel jobs, which is not an uncommon number for iOS/Android development where many combinations of app, device and OS version need to be tested. We'd actually like to split some of the jobs into smaller chunks to take advantage of idle build agents, but this would <b>greatly</b> exacerbate the problem.</p>

<p>Grouping jobs under multiple stages would improve the UI experience, but also drastically increase the runtime of our integration runs as stages are executed serially.</p>

<p>I envision two possible solutions:</p>

<p>1. Stages have a 'parallel' option that allows them to run at the same time as other parallel stages.<br/>
2. A step is introduced that is used purely as an annotation for the purposes of rendering a more appropriate graph. Ideally the step would be deeply nestable allowing for complex graph hierarchies.</p>

<p>Thanks for all the hard work on Blue Ocean, it's really shaping up nicely and I eagerly await each new release.</p>

---
<details><summary><i>Originally reported by <img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/ileitch.jpg" title="ileitch's avatar" /> <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=ileitch">ileitch</a>, imported from: <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-41205" target="_blank">Stage graph unsuitable for large and/or complex pipelines</a></i></summary>
<i><ul>
<li><b>assignee</b>: <img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/benlangfeld.png" title="benlangfeld's avatar" /> <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=benlangfeld">benlangfeld</a>
<li><b>status</b>: Resolved
<li><b>priority</b>: Major
<li><b>component(s)</b>: blueocean-plugin
<li><b>resolution</b>: Fixed
<li><b>resolved</b>: 2017-11-16T23:07:45+00:00
<li><b>votes</b>: 19
<li><b>watchers</b>: 36
<li><b>imported</b>: 2025-12-01
</ul></i>
<details><summary>Raw content of original issue</summary>

<pre>
<div class="panel" style="border-width: 1px;"><div class="panelHeader" style="border-bottom-width: 1px;background-color: #3878de;color: white;"><b>Improvement on roadmap</b></div><div class="panelContent">
<p>This improvement is on the Blue Ocean project roadmap. Check the <a href="https://jenkins.io/projects/blueocean/roadmap/" class="external-link" target="_blank" rel="nofollow noopener">roadmap page</a> for updates. </p>
</div></div>

<p>The Blue Ocean stage graph is great for small, simple pipelines however it breaks down with many parallel builds. See attached screenshot for an example.</p>

<p>Because 'stage' can no longer be nested within 'parallel', all of our steps must belong under a single 'Test' stage. We have 19 parallel jobs, which is not an uncommon number for iOS/Android development where many combinations of app, device and OS version need to be tested. We'd actually like to split some of the jobs into smaller chunks to take advantage of idle build agents, but this would <b>greatly</b> exacerbate the problem.</p>

<p>Grouping jobs under multiple stages would improve the UI experience, but also drastically increase the runtime of our integration runs as stages are executed serially.</p>

<p>I envision two possible solutions:</p>

<p>1. Stages have a 'parallel' option that allows them to run at the same time as other parallel stages.
2. A step is introduced that is used purely as an annotation for the purposes of rendering a more appropriate graph. Ideally the step would be deeply nestable allowing for complex graph hierarchies.</p>

<p>Thanks for all the hard work on Blue Ocean, it's really shaping up nicely and I eagerly await each new release.</p></pre>
</details>
</details>
<details><summary><i>environment</i></summary>

```
Jenkins 2.40<br/>
Blue Ocean 1.0.0-b17
```
</details>
<details><summary><i>3 attachments</i></summary>

- [Screen Shot 2017-01-19 at 7.20.51 PM.png](https://raw.githubusercontent.com/jenkinsci/attachments-from-jira-issues-misc/refs/heads/main/attachments/35527/Screen%20Shot%202017-01-19%20at%207.20.51%20PM.png)
  > ![Screen Shot 2017-01-19 at 7.20.51 PM.png](https://raw.githubusercontent.com/jenkinsci/attachments-from-jira-issues-misc/refs/heads/main/attachments/35527/Screen%20Shot%202017-01-19%20at%207.20.51%20PM.png)
- [matrix-test-parallel.jpg](https://raw.githubusercontent.com/jenkinsci/attachments-from-jira-issues-misc/refs/heads/main/attachments/35767/matrix-test-parallel.jpg)
  > ![matrix-test-parallel.jpg](https://raw.githubusercontent.com/jenkinsci/attachments-from-jira-issues-misc/refs/heads/main/attachments/35767/matrix-test-parallel.jpg)
- [too-much-pipeline.png](https://raw.githubusercontent.com/jenkinsci/attachments-from-jira-issues-misc/refs/heads/main/attachments/36852/too-much-pipeline.png)
  > ![too-much-pipeline.png](https://raw.githubusercontent.com/jenkinsci/attachments-from-jira-issues-misc/refs/heads/main/attachments/36852/too-much-pipeline.png)
</details>

**Comment by jenkins-infra-bot** [comment]
- _Remote links associated with this issue:_


  - [CloudBees Internal UX-624](https://cloudbees.atlassian.net/browse/UX-624)
  - [Page](https://wiki.jenkins.io/pages/viewpage.action?pageId=98403475)

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/jamesdumay.png" title="jamesdumay's avatar" /> jamesdumay:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-41205?focusedId=284154&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-284154">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Thanks <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=ileitch" class="user-hover" rel="ileitch">ileitch</a>! For that second point would you be interested in something like <a href="https://issues.jenkins.io/browse/JENKINS-38442" title="View sequential stages in the pipeline visualization graph" class="issue-link" data-issue-key="JENKINS-38442"><del>JENKINS-38442</del></a> ?</p>
</pre>
</details>

</ul>
</details>

<p>Thanks <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=ileitch" class="user-hover" rel="ileitch">ileitch</a>! For that second point would you be interested in something like <a href="https://issues.jenkins.io/browse/JENKINS-38442" title="View sequential stages in the pipeline visualization graph" class="issue-link" data-issue-key="JENKINS-38442"><del>JENKINS-38442</del></a> ?</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/ileitch.jpg" title="ileitch's avatar" /> ileitch:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-41205?focusedId=284160&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-284160">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=jamesdumay" class="user-hover" rel="jamesdumay">jamesdumay</a> It's not clear in <a href="https://issues.jenkins.io/browse/JENKINS-38442" title="View sequential stages in the pipeline visualization graph" class="issue-link" data-issue-key="JENKINS-38442"><del>JENKINS-38442</del></a> how the graph would be rendered. Nested stages would solve the issue for me provided they do alter the graph, however ideally both nested stages and parallel stages would provide the most flexibility.</p>
</pre>
</details>

</ul>
</details>

<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=jamesdumay" class="user-hover" rel="jamesdumay">jamesdumay</a> It's not clear in <a href="https://issues.jenkins.io/browse/JENKINS-38442" title="View sequential stages in the pipeline visualization graph" class="issue-link" data-issue-key="JENKINS-38442"><del>JENKINS-38442</del></a> how the graph would be rendered. Nested stages would solve the issue for me provided they do alter the graph, however ideally both nested stages and parallel stages would provide the most flexibility.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/jamesdumay.png" title="jamesdumay's avatar" /> jamesdumay:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-41205?focusedId=284161&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-284161">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=ileitch" class="user-hover" rel="ileitch">ileitch</a> we are not even clear how things should be rendered because users ideas of how that should look are so diverse <img class="emoticon" src="https://issues.jenkins.io/images/icons/emoticons/smile.png" height="16" width="16" align="absmiddle" alt="" border="0"/></p>

<p>I think there are some things we can do here to make the existing graph more acceptable in your case. We will also be looking at how to solve the nesting problem this year but I can't give a good ETA on it yet.</p>
</pre>
</details>

</ul>
</details>

<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=ileitch" class="user-hover" rel="ileitch">ileitch</a> we are not even clear how things should be rendered because users ideas of how that should look are so diverse <img class="emoticon" src="https://issues.jenkins.io/images/icons/emoticons/smile.png" height="16" width="16" align="absmiddle" alt="" border="0"/></p>

<p>I think there are some things we can do here to make the existing graph more acceptable in your case. We will also be looking at how to solve the nesting problem this year but I can't give a good ETA on it yet.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/ileitch.jpg" title="ileitch's avatar" /> ileitch:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-41205?focusedId=284162&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-284162">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>The sketch you provided: <a href="https://issues.jenkins-ci.org/secure/attachment/34100/blueocean.sketch%202016-09-28%2015-03-57.png" class="external-link" target="_blank" rel="nofollow noopener">https://issues.jenkins-ci.org/secure/attachment/34100/blueocean.sketch%202016-09-28%2015-03-57.png</a>
 would not be ideal as it only alters the steps table, not the graph. If we applied this approach to our pipeline, the graph would only display a single "Test" stage, with all of the meaningful stages being shown in the list below. This would make the graph linear, and somewhat redundant.</p>
</pre>
</details>

</ul>
</details>

<p>The sketch you provided: <a href="https://issues.jenkins-ci.org/secure/attachment/34100/blueocean.sketch%202016-09-28%2015-03-57.png" class="external-link" target="_blank" rel="nofollow noopener">https://issues.jenkins-ci.org/secure/attachment/34100/blueocean.sketch%202016-09-28%2015-03-57.png</a><br/>
 would not be ideal as it only alters the steps table, not the graph. If we applied this approach to our pipeline, the graph would only display a single "Test" stage, with all of the meaningful stages being shown in the list below. This would make the graph linear, and somewhat redundant.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/jamesdumay.png" title="jamesdumay's avatar" /> jamesdumay:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-41205?focusedId=284163&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-284163">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=ileitch" class="user-hover" rel="ileitch">ileitch</a> yeah I think that design is totally out for the reason you provided and because two steps can appear on the screen it makes the running behaviour really complex.</p>
</pre>
</details>

</ul>
</details>

<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=ileitch" class="user-hover" rel="ileitch">ileitch</a> yeah I think that design is totally out for the reason you provided and because two steps can appear on the screen it makes the running behaviour really complex.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/jamesdumay.png" title="jamesdumay's avatar" /> jamesdumay:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-41205?focusedId=284164&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-284164">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=ileitch" class="user-hover" rel="ileitch">ileitch</a> actually would you be up for a Google Hangout in the next few weeks? We can discuss this in greater detail. Would love to see some of the Pipelines you have built and how they show up in Blue Ocean.</p>
</pre>
</details>

</ul>
</details>

<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=ileitch" class="user-hover" rel="ileitch">ileitch</a> actually would you be up for a Google Hangout in the next few weeks? We can discuss this in greater detail. Would love to see some of the Pipelines you have built and how they show up in Blue Ocean.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/ileitch.jpg" title="ileitch's avatar" /> ileitch:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-41205?focusedId=284176&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-284176">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=jamesdumay" class="user-hover" rel="jamesdumay">jamesdumay</a> Please see my proposal here: <a href="https://issues.jenkins-ci.org/browse/JENKINS-38442?focusedCommentId=284175&amp;page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-284175" class="external-link" target="_blank" rel="nofollow noopener">https://issues.jenkins-ci.org/browse/JENKINS-38442?focusedCommentId=284175&amp;page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-284175</a></p>
</pre>
</details>

</ul>
</details>

<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=jamesdumay" class="user-hover" rel="jamesdumay">jamesdumay</a> Please see my proposal here: <a href="https://issues.jenkins-ci.org/browse/JENKINS-38442?focusedCommentId=284175&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-284175" class="external-link" target="_blank" rel="nofollow noopener">https://issues.jenkins-ci.org/browse/JENKINS-38442?focusedCommentId=284175&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-284175</a></p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/ileitch.jpg" title="ileitch's avatar" /> ileitch:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-41205?focusedId=284177&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-284177">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=jamesdumay" class="user-hover" rel="jamesdumay">jamesdumay</a> Sorry just noticed the Google Hangout question. Yeah I'm happy to do so, though be warned that I'm very new to Jenkins (I was tasked with switching away from TeamCity). We only have a single Pipeline at the moment.</p>
</pre>
</details>

</ul>
</details>

<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=jamesdumay" class="user-hover" rel="jamesdumay">jamesdumay</a> Sorry just noticed the Google Hangout question. Yeah I'm happy to do so, though be warned that I'm very new to Jenkins (I was tasked with switching away from TeamCity). We only have a single Pipeline at the moment.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/jamesdumay.png" title="jamesdumay's avatar" /> jamesdumay:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-41205?focusedId=284179&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-284179">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Oh no that's perfect that you are new to things! Can you email me on your work email? I'm jdumay@cloudbees.com</p>
</pre>
</details>

</ul>
</details>

<p>Oh no that's perfect that you are new to things! Can you email me on your work email? I'm jdumay@​cloudbees.com</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>evildeece:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-41205?focusedId=288778&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-288778">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>A multiaxis way of displaying steps would be nice too, consider the following:</p>


<div class="code panel" style="border-width: 1px;"><div class="codeContent panelContent">
<pre class="code-java">def hostPlatforms = [<span class="code-quote">'ppc64le'</span>, <span class="code-quote">'x86_64'</span>]
def targetPlatforms = [<span class="code-quote">'x86_64'</span>, <span class="code-quote">'power32'</span>, <span class="code-quote">'ppc64le'</span>, <span class="code-quote">'ppc64be'</span>, <span class="code-quote">'arm'</span>]
def hostOSs = [<span class="code-quote">'ubuntu-16.04'</span>]

def builds = [:]

<span class="code-keyword">for</span> (hostPlatform in hostPlatforms) {
    <span class="code-keyword">for</span> (targetPlatform in targetPlatforms) {
        <span class="code-keyword">if</span> (hostPlatform.equals(<span class="code-quote">'ppc64le'</span>) &amp;&amp; targetPlatform.equals(<span class="code-quote">'x86_64'</span>)) {
            <span class="code-keyword">continue</span>;
        }

        <span class="code-keyword">for</span> (hostOS in hostOSs) {
            def name = <span class="code-quote">'build-'</span> + hostPlatform + <span class="code-quote">'-'</span> + targetPlatform
            def myHostPlatform = hostPlatform
            def myHostOS = hostOS
            def myTargetPlatform = targetPlatform

            builds[name] = {buildToolchain(myHostPlatform, myHostOS, myTargetPlatform)}
        }
    }
}

stage(<span class="code-quote">'build'</span>) {
    parallel builds
}
</pre>
</div></div>

<p>This currently ends up with a long list scrolling down the page, but would be better represented by a display similar to the Matrix Project plugin.</p>
</pre>
</details>

</ul>
</details>

<p>A multiaxis way of displaying steps would be nice too, consider the following:</p>



<pre>
def hostPlatforms = [<span class="code-quote">'ppc64le'</span>, <span class="code-quote">'x86_64'</span>]
def targetPlatforms = [<span class="code-quote">'x86_64'</span>, <span class="code-quote">'power32'</span>, <span class="code-quote">'ppc64le'</span>, <span class="code-quote">'ppc64be'</span>, <span class="code-quote">'arm'</span>]
def hostOSs = [<span class="code-quote">'ubuntu-16.04'</span>]

def builds = [:]

<span class="code-keyword">for</span> (hostPlatform in hostPlatforms) {
    <span class="code-keyword">for</span> (targetPlatform in targetPlatforms) {
<span class="code-keyword">if</span> (hostPlatform.equals(<span class="code-quote">'ppc64le'</span>) && targetPlatform.equals(<span class="code-quote">'x86_64'</span>)) {
    <span class="code-keyword">continue</span>;
}

<span class="code-keyword">for</span> (hostOS in hostOSs) {
    def name = <span class="code-quote">'build-'</span> + hostPlatform + <span class="code-quote">'-'</span> + targetPlatform
    def myHostPlatform = hostPlatform
    def myHostOS = hostOS
    def myTargetPlatform = targetPlatform

    builds[name] = {buildToolchain(myHostPlatform, myHostOS, myTargetPlatform)}
}
    }
}

stage(<span class="code-quote">'build'</span>) {
    parallel builds
}
</pre>

<p>This currently ends up with a long list scrolling down the page, but would be better represented by a display similar to the Matrix Project plugin.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>scgruber:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-41205?focusedId=294080&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-294080">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p><span class="image-wrap" style=""><a id="36852_thumb" href="https://issues.jenkins.io/secure/attachment/36852/36852_too-much-pipeline.png" title="too-much-pipeline.png" file-preview-type="image" file-preview-id="36852" file-preview-title="too-much-pipeline.png"><img src="https://issues.jenkins.io/secure/thumbnail/36852/_thumb_36852.png" style="border: 0px solid black" role="presentation"/></a></span></p>

<p>This pipeline of ours is unworkable in the current view (1.0.0rc3) because we cannot click through to any information on our later stages (we have approximately 250 stages total, and this view only shows the first 90). Even if the full graph is unrenderable, it would be nice to have some way of loading these stages. Just a dropdown list would be preferable to the current state.</p>
</pre>
</details>

</ul>
</details>

<p><span class="image-wrap" style=""><a id="36852_thumb" href="https://raw.githubusercontent.com/jenkinsci/attachments-from-jira-issues-misc/refs/heads/main/attachments/36852/too-much-pipeline.png" title="too-much-pipeline.png" file-preview-type="image" file-preview-id="36852" file-preview-title="too-much-pipeline.png"><img src="https://raw.githubusercontent.com/jenkinsci/attachments-from-jira-issues-misc/refs/heads/main/attachments/36852/too-much-pipeline.png" style="border: 0px solid black" role="presentation"/></a></span></p>

<p>This pipeline of ours is unworkable in the current view (1.0.0rc3) because we cannot click through to any information on our later stages (we have approximately 250 stages total, and this view only shows the first 90). Even if the full graph is unrenderable, it would be nice to have some way of loading these stages. Just a dropdown list would be preferable to the current state.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/jamesdumay.png" title="jamesdumay's avatar" /> jamesdumay:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-41205?focusedId=294109&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-294109">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=scgruber" class="user-hover" rel="scgruber">scgruber</a> great description of your use case here. I have a few questions:</p>
<ul>
	<li>What actions are you performing with the parallels?</li>
	<li>Why 250 stages? What meaning do they have?</li>
</ul>

</pre>
</details>

</ul>
</details>

<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=scgruber" class="user-hover" rel="scgruber">scgruber</a> great description of your use case here. I have a few questions:</p>
<ul>
	<li>What actions are you performing with the parallels?</li>
	<li>Why 250 stages? What meaning do they have?</li>
</ul>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>scgruber:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-41205?focusedId=294195&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-294195">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=jamesdumay" class="user-hover" rel="jamesdumay">jamesdumay</a> this pipeline is an integration test against our different hardware platforms. We have a sequence of tests (asdf3, asdf4, ...) which run in parallel over our hardware platforms (zxcv1, zxcv2, zxcv3, ...). So we have a stage for each combination of test and hardware, and that's how we get to 250.</p>
</pre>
</details>

</ul>
</details>

<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=jamesdumay" class="user-hover" rel="jamesdumay">jamesdumay</a> this pipeline is an integration test against our different hardware platforms. We have a sequence of tests (asdf3, asdf4, ...) which run in parallel over our hardware platforms (zxcv1, zxcv2, zxcv3, ...). So we have a stage for each combination of test and hardware, and that's how we get to 250.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/jamesdumay.png" title="jamesdumay's avatar" /> jamesdumay:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-41205?focusedId=294252&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-294252">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=scgruber" class="user-hover" rel="scgruber">scgruber</a> very interesting! Thanks for the additional context. When we plan this feature, could we contact you for a Google Hangout to discuss the new design?</p>
</pre>
</details>

</ul>
</details>

<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=scgruber" class="user-hover" rel="scgruber">scgruber</a> very interesting! Thanks for the additional context. When we plan this feature, could we contact you for a Google Hangout to discuss the new design?</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>scgruber:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-41205?focusedId=294330&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-294330">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=jdumay" class="user-hover" rel="jdumay">jdumay</a> sure, as long as it can be scheduled during US Pacific (UTC-7) business hours.</p>
</pre>
</details>

</ul>
</details>

<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=jdumay" class="user-hover" rel="jdumay">jdumay</a> sure, as long as it can be scheduled during US Pacific (UTC-7) business hours.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/clementgautier.jpg" title="clementgautier's avatar" /> clementgautier:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-41205?focusedId=299222&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-299222">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Any insight on this issue ? I'm in the same situation with even more parallel steps (over 400 and growing). Basically we have so much browser tests that we need to split them a lot. Previously we were using a distributed system of our own using a queue system and consumers that upload the results back to the Jenkins master server but it kinda suck and we really want to move back to a basic master / slave management of your Jenkins.</p>

<p>How can I help ? From my point of view, if we could simply say, using parallel, to not create a visual step for each branch of the parallel would be enough as a quickfix. Do you have any slack or whatever tool to talk directly ? I'm not used to Java so I might need some help with the contribution process.</p>
</pre>
</details>

</ul>
</details>

<p>Any insight on this issue ? I'm in the same situation with even more parallel steps (over 400 and growing). Basically we have so much browser tests that we need to split them a lot. Previously we were using a distributed system of our own using a queue system and consumers that upload the results back to the Jenkins master server but it kinda suck and we really want to move back to a basic master / slave management of your Jenkins.</p>

<p>How can I help ? From my point of view, if we could simply say, using parallel, to not create a visual step for each branch of the parallel would be enough as a quickfix. Do you have any slack or whatever tool to talk directly ? I'm not used to Java so I might need some help with the contribution process.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/andrewconti.png" title="andrewconti's avatar" /> andrewconti:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-41205?focusedId=302862&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-302862">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>I'm in a similar situation. I've got 25 separate build combinations, some of which are virtualized and have one or more packaging nodes as well. Each artifact produced then gets a test node, and all told there's over 40 artifacts to test. All told I've got about 100 nodes so I'm guessing that must be the limit. Just my 2¢ that the limit is too low.</p>
</pre>
</details>

</ul>
</details>

<p>I'm in a similar situation. I've got 25 separate build combinations, some of which are virtualized and have one or more packaging nodes as well. Each artifact produced then gets a test node, and all told there's over 40 artifacts to test. All told I've got about 100 nodes so I'm guessing that must be the limit. Just my 2¢ that the limit is too low.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>philster_jenkins:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-41205?focusedId=304844&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-304844">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>+1 on the limit of 100 being too low.  I have some pipelines that are slightly larger than 100, and this is a real pain.  My pipelines only have a few (non-nested) stages, but have lots of parallel steps per stage.  I'm fine with dealing with a slightly long UI graph if that means I get to see all the nodes.</p>

<p> </p>

<p>Is there any chance that a short-term solution could be put in place to allow the limit to be configurable, with no changes to the current UI graph display?  I have a feeling that refactoring the UI using some of the solutions proposed here will take quite a while, especially since the roadmap indicates it is "Not planned.</p>

<p> </p>

<p>I looked into it a bit.  </p>

<p>The current UI makes one call to:</p>
<div class="preformatted panel" style="border-width: 1px;"><div class="preformattedContent panelContent">
<pre>http://jenkins/blue/rest/organizations/jenkins/pipelines/_job_/runs/_runId_/nodes/
</pre>
</div></div>
<p>Internally, the nodes are retrieved via the <a href="https://github.com/jenkinsci/blueocean-plugin/blob/a8d2683497244bc4fe247550353b39a979f3989a/blueocean-pipeline-api-impl/src/main/java/io/jenkins/blueocean/rest/impl/pipeline/PipelineNodeContainerImpl.java#L42" class="external-link" target="_blank" rel="nofollow noopener">PipelineNodeContainerImpl</a>, which is a <a href="https://github.com/jenkinsci/blueocean-plugin/blob/8360d58c84fe60faf9d0867fa638d37827402aad/blueocean-rest/src/main/java/io/jenkins/blueocean/rest/model/BluePipelineNodeContainer.java" class="external-link" target="_blank" rel="nofollow noopener">BluePipelineNodeContainer</a>, which is a <a href="https://github.com/jenkinsci/blueocean-plugin/blob/64b1fd9b510d0e7cc32995d5a7d3a97345354f65/blueocean-rest/src/main/java/io/jenkins/blueocean/rest/model/Container.java" class="external-link" target="_blank" rel="nofollow noopener">Container</a>, which is a <a href="https://github.com/jenkinsci/blueocean-plugin/blob/99cbbdc586c0d1e888ac95d7f2daebe2d6254436/blueocean-rest/src/main/java/io/jenkins/blueocean/rest/pageable/Pageable.java" class="external-link" target="_blank" rel="nofollow noopener">Pageable</a>.</p>

<p>Therefore, the response that is constructed is a <a href="https://github.com/jenkinsci/blueocean-plugin/blob/dc86ece5e36c2008c096883baf2594f88efe8bd6/blueocean-rest/src/main/java/io/jenkins/blueocean/rest/pageable/PagedResponse.java" class="external-link" target="_blank" rel="nofollow noopener">PagedResponse</a>, which has a <a href="https://github.com/jenkinsci/blueocean-plugin/blob/dc86ece5e36c2008c096883baf2594f88efe8bd6/blueocean-rest/src/main/java/io/jenkins/blueocean/rest/pageable/PagedResponse.java#L32" class="external-link" target="_blank" rel="nofollow noopener">default limit of 100</a>.  Since no start/limit is passed in the request, the default start=0 and limit=100 are used.  So even though the container has all of the nodes in it, only the first 100 are returned in the HTTP response.</p>

<p>The response also includes a Link header for the next page:</p>
<div class="preformatted panel" style="border-width: 1px;"><div class="preformattedContent panelContent">
<pre>Link: &lt;/blue/rest/organizations/jenkins/pipelines/_job_/runs/_runId_/nodes/?start=100&amp;limit=100&gt;; rel="next"
</pre>
</div></div>
<p>However, the UI does not follow the link to retrieve the next page.</p>

<p> </p>

<p>While debugging, I was able to set a breakpoint on the server side, and hack around the arbitrary 100 limit to make the server return more than 100.  The UI worked just fine with the number of nodes in my pipelines.</p>

<p> </p>

<p>There are a few hacky ways to make all the nodes show on the current UI:</p>
<ol>
	<li>Increase the default limit constant</li>
	<li>Provide a way to configure the default limit (and don't rely on a constant), therefore allowing users to modify the limit as they desire</li>
	<li>Have the call to .../nodes pass a higher limit by default... e.g. <tt>.../nodes/?limit=500</tt></li>
	<li>Have the UI follow links to next pages (could be done by default, OR done if the user clicks on the "Unable to display more" node.</li>
</ol>


<p> </p>

<p>In any case, a short term solution would be much appreciated.</p>

<p> </p>

<p> </p>
</pre>
</details>

</ul>
</details>

<p>+1 on the limit of 100 being too low.  I have some pipelines that are slightly larger than 100, and this is a real pain.  My pipelines only have a few (non-nested) stages, but have lots of parallel steps per stage.  I'm fine with dealing with a slightly long UI graph if that means I get to see all the nodes.</p>

<p> </p>

<p>Is there any chance that a short-term solution could be put in place to allow the limit to be configurable, with no changes to the current UI graph display?  I have a feeling that refactoring the UI using some of the solutions proposed here will take quite a while, especially since the roadmap indicates it is "Not planned.</p>

<p> </p>

<p>I looked into it a bit.  </p>

<p>The current UI makes one call to:</p>


```
http://jenkins/blue/rest/organizations/jenkins/pipelines/_job_/runs/_runId_/nodes/

```
<p>Internally, the nodes are retrieved via the <a href="https://github.com/jenkinsci/blueocean-plugin/blob/a8d2683497244bc4fe247550353b39a979f3989a/blueocean-pipeline-api-impl/src/main/java/io/jenkins/blueocean/rest/impl/pipeline/PipelineNodeContainerImpl.java#L42" class="external-link" target="_blank" rel="nofollow noopener">PipelineNodeContainerImpl</a>, which is a <a href="https://github.com/jenkinsci/blueocean-plugin/blob/8360d58c84fe60faf9d0867fa638d37827402aad/blueocean-rest/src/main/java/io/jenkins/blueocean/rest/model/BluePipelineNodeContainer.java" class="external-link" target="_blank" rel="nofollow noopener">BluePipelineNodeContainer</a>, which is a <a href="https://github.com/jenkinsci/blueocean-plugin/blob/64b1fd9b510d0e7cc32995d5a7d3a97345354f65/blueocean-rest/src/main/java/io/jenkins/blueocean/rest/model/Container.java" class="external-link" target="_blank" rel="nofollow noopener">Container</a>, which is a <a href="https://github.com/jenkinsci/blueocean-plugin/blob/99cbbdc586c0d1e888ac95d7f2daebe2d6254436/blueocean-rest/src/main/java/io/jenkins/blueocean/rest/pageable/Pageable.java" class="external-link" target="_blank" rel="nofollow noopener">Pageable</a>.</p>

<p>Therefore, the response that is constructed is a <a href="https://github.com/jenkinsci/blueocean-plugin/blob/dc86ece5e36c2008c096883baf2594f88efe8bd6/blueocean-rest/src/main/java/io/jenkins/blueocean/rest/pageable/PagedResponse.java" class="external-link" target="_blank" rel="nofollow noopener">PagedResponse</a>, which has a <a href="https://github.com/jenkinsci/blueocean-plugin/blob/dc86ece5e36c2008c096883baf2594f88efe8bd6/blueocean-rest/src/main/java/io/jenkins/blueocean/rest/pageable/PagedResponse.java#L32" class="external-link" target="_blank" rel="nofollow noopener">default limit of 100</a>.  Since no start/limit is passed in the request, the default start=0 and limit=100 are used.  So even though the container has all of the nodes in it, only the first 100 are returned in the HTTP response.</p>

<p>The response also includes a Link header for the next page:</p>


```
Link: </blue/rest/organizations/jenkins/pipelines/_job_/runs/_runId_/nodes/?start=100&limit=100>; rel="next"

```
<p>However, the UI does not follow the link to retrieve the next page.</p>

<p> </p>

<p>While debugging, I was able to set a breakpoint on the server side, and hack around the arbitrary 100 limit to make the server return more than 100.  The UI worked just fine with the number of nodes in my pipelines.</p>

<p> </p>

<p>There are a few hacky ways to make all the nodes show on the current UI:</p>
<ol>
	<li>Increase the default limit constant</li>
	<li>Provide a way to configure the default limit (and don't rely on a constant), therefore allowing users to modify the limit as they desire</li>
	<li>Have the call to .../nodes pass a higher limit by default... e.g. <tt>.../nodes/?limit=500</tt></li>
	<li>Have the UI follow links to next pages (could be done by default, OR done if the user clicks on the "Unable to display more" node.</li>
</ol>


<p> </p>

<p>In any case, a short term solution would be much appreciated.</p>

<p> </p>

<p> </p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/benlangfeld.png" title="benlangfeld's avatar" /> benlangfeld:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-41205?focusedId=312492&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-312492">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>My team has the necessary skills to contribute a fix to the UI to fetch all available pages. If we were to submit such a PR, is there a reasonable chance it would be merged for inclusion in the next BlueOcean release, or is there more roadmap politics to navigate than that?</p>

<p>For anyone who is curious, the deficient object is the PipelinePager which doesn't actually page in data: <a href="https://github.com/jenkinsci/blueocean-plugin/blob/master/blueocean-dashboard/src/main/js/components/karaoke/services/pagers/PipelinePager.js" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/jenkinsci/blueocean-plugin/blob/master/blueocean-dashboard/src/main/js/components/karaoke/services/pagers/PipelinePager.js</a> past the initial page.</p>
</pre>
</details>

</ul>
</details>

<p>My team has the necessary skills to contribute a fix to the UI to fetch all available pages. If we were to submit such a PR, is there a reasonable chance it would be merged for inclusion in the next BlueOcean release, or is there more roadmap politics to navigate than that?</p>

<p>For anyone who is curious, the deficient object is the PipelinePager which doesn't actually page in data: <a href="https://github.com/jenkinsci/blueocean-plugin/blob/master/blueocean-dashboard/src/main/js/components/karaoke/services/pagers/PipelinePager.js" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/jenkinsci/blueocean-plugin/blob/master/blueocean-dashboard/src/main/js/components/karaoke/services/pagers/PipelinePager.js</a> past the initial page.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>garettarrowood:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-41205?focusedId=315885&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-315885">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>This bug it greatly hindering my work. Could it be moved up the roadmap or could someone respond to the gentleman that offered to fix it in the above comment?</p>
</pre>
</details>

</ul>
</details>

<p>This bug it greatly hindering my work. Could it be moved up the roadmap or could someone respond to the gentleman that offered to fix it in the above comment?</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/kzantow.png" title="kzantow's avatar" /> kzantow:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-41205?focusedId=317239&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-317239">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=benlangfeld" class="user-hover" rel="benlangfeld">benlangfeld</a> if you submit a PR that fixes the issue it would absolutely be considered for inclusion; we'd just have a look and make sure tests pass, etc.. Submissions are <em>always</em> welcome!</p>
</pre>
</details>

</ul>
</details>

<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=benlangfeld" class="user-hover" rel="benlangfeld">benlangfeld</a> if you submit a PR that fixes the issue it would absolutely be considered for inclusion; we'd just have a look and make sure tests pass, etc.. Submissions are <em>always</em> welcome!</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/cliffmeyers.png" title="cliffmeyers's avatar" /> cliffmeyers:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-41205?focusedId=317240&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-317240">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Seconded. <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=benlangfeld" class="user-hover" rel="benlangfeld">benlangfeld</a> I had looked at this problem in the past. Another option is to do successive fetches until all nodes / stages are loaded. If you look at the REST responses, you'll see there is pagination data written into a "Link" response header IIRC. That's a way to determine whether there is additional data to be fetched, and you could write some logic to grab say n=100 and just perform successive fetches until the Link header indicates there is no more data.</p>

<p>We may want to be careful about doing a massive fetch up front (say n=500) as for complex pipelines this might have a perf impact server wide. I recall discussing this with <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=vivek" class="user-hover" rel="vivek">vivek</a> a while back, can you refresh my memory on whether it might be preferable to do a single large fetch (say n=500) or several smaller fetches (n=100) until all data is loaded? Intuitively fewer large fetches seems more efficient from client's perspective, but I seem to recall a concern with loading a large number of nodes concurrently in the context of a single request?</p>
</pre>
</details>

</ul>
</details>

<p>Seconded. <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=benlangfeld" class="user-hover" rel="benlangfeld">benlangfeld</a> I had looked at this problem in the past. Another option is to do successive fetches until all nodes / stages are loaded. If you look at the REST responses, you'll see there is pagination data written into a "Link" response header IIRC. That's a way to determine whether there is additional data to be fetched, and you could write some logic to grab say n=100 and just perform successive fetches until the Link header indicates there is no more data.</p>

<p>We may want to be careful about doing a massive fetch up front (say n=500) as for complex pipelines this might have a perf impact server wide. I recall discussing this with <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=vivek" class="user-hover" rel="vivek">vivek</a> a while back, can you refresh my memory on whether it might be preferable to do a single large fetch (say n=500) or several smaller fetches (n=100) until all data is loaded? Intuitively fewer large fetches seems more efficient from client's perspective, but I seem to recall a concern with loading a large number of nodes concurrently in the context of a single request?</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/benlangfeld.png" title="benlangfeld's avatar" /> benlangfeld:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-41205?focusedId=317246&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-317246">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Following Link is precisely what I had in mind <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=cliffmeyers" class="user-hover" rel="cliffmeyers">cliffmeyers</a>. We'll prep a patch. Thanks everyone.</p>
</pre>
</details>

</ul>
</details>

<p>Following Link is precisely what I had in mind <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=cliffmeyers" class="user-hover" rel="cliffmeyers">cliffmeyers</a>. We'll prep a patch. Thanks everyone.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/michaelneale.png" title="michaelneale's avatar" /> michaelneale:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-41205?focusedId=317283&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-317283">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=benlangfeld" class="user-hover" rel="benlangfeld">benlangfeld</a> just make sure you are nice and up to date with master as some recent changes were merged for how it follows along (may not affect you, just FYI). As for fetching the pages - absolutely why not, if you have a PR that would be wonderful. Go for it!</p>
</pre>
</details>

</ul>
</details>

<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=benlangfeld" class="user-hover" rel="benlangfeld">benlangfeld</a> just make sure you are nice and up to date with master as some recent changes were merged for how it follows along (may not affect you, just FYI). As for fetching the pages - absolutely why not, if you have a PR that would be wonderful. Go for it!</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/benlangfeld.png" title="benlangfeld's avatar" /> benlangfeld:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-41205?focusedId=318135&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-318135">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>A patch to resolve this is proposed at <a href="https://github.com/jenkinsci/blueocean-plugin/pull/1517." class="external-link" target="_blank" rel="nofollow noopener">https://github.com/jenkinsci/blueocean-plugin/pull/1517.</a> I would appreciate a review, particularly from <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=cliffmeyers" class="user-hover" rel="cliffmeyers">cliffmeyers</a>.</p>
</pre>
</details>

</ul>
</details>

<p>A patch to resolve this is proposed at <a href="https://github.com/jenkinsci/blueocean-plugin/pull/1517." class="external-link" target="_blank" rel="nofollow noopener">https://github.com/jenkinsci/blueocean-plugin/pull/1517.</a> I would appreciate a review, particularly from <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=cliffmeyers" class="user-hover" rel="cliffmeyers">cliffmeyers</a>.</p>

**Comment by jenkins-infra-bot** [comment]
<i>[Original `depends_on` from Jira: <a href="https://github.com/jenkinsci/blueocean-plugin/issues?q=is%3Aissue%20%22jira_issue_key%3DJENKINS-47683%22">JENKINS-47683</a>]</i>

**Comment by jenkins-infra-bot** [comment]
<i>[Original `blocks` from Jira: <a href="https://github.com/jenkinsci/blueocean-plugin/issues?q=is%3Aissue%20%22jira_issue_key%3DJENKINS-39770%22">JENKINS-39770</a>]</i>

**Comment by jenkins-infra-bot** [comment]
<i>[Original `epic` from Jira: <a href="https://github.com/jenkinsci/blueocean-plugin/issues?q=is%3Aissue%20%22jira_issue_key%3DJENKINS-43953%22">JENKINS-43953</a>]</i>
